### PR TITLE
support retrieving helm charts from private helm registry

### PR DESCRIPTION
**Pull request by cdoron**
We use a mechanism similar to that used by kubernetes code to pull docker images:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/kuberuntime/kuberuntime_image.go#L31

Following Roee's suggestion, we add an explicit field for the secret next to the chart field in FybrikModule. That way you would know that login is required and what secret to use.

We use https://github.com/vdemeester/k8s-pkg-credentialprovider
to overcome the problem with using `k8s.io/kubernetes` as a dependency

### CODE REVIEW

File: manager/controllers/app/blueprint_controller.go [hunk 1]
```diff
@@ -141,6 +155,42 @@ func (r *BlueprintReconciler) applyChartResource(log logr.Logger, chartSpec app.
 	nbytes, _ := yaml.Marshal(args)
 	log.Info(fmt.Sprintf("--- Values.yaml ---\n\n%s\n\n", nbytes))
 
+	if chartSpec.ChartPullSecret != "" {
+		// obtain ChartPullSecret
+		pullSecret := corev1.Secret{}
+		pullSecrets := []corev1.Secret{}
+
+		if err := r.Get(ctx, types.NamespacedName{Namespace: utils.GetSystemNamespace(), Name: chartSpec.ChartPullSecret},
+			&pullSecret); err == nil {
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:164]: Do you have RBAC for secrets in system namespace? Is corev1.Secret mapped to the system namespace in manager cache?

Comment (cdoron) [reply manager/controllers/app/blueprint_controller.go:164]: hi @shlomitk1 : thank you for helping me confirm that we have RBAC for secrets in system namespace

File: manager/controllers/app/blueprint_controller.go [hunk 2]
```diff
@@ -141,6 +155,42 @@ func (r *BlueprintReconciler) applyChartResource(log logr.Logger, chartSpec app.
...
+		keyring, _ = credentialprovidersecrets.MakeDockerKeyring(pullSecrets, keyring)
+		repoToPull, err := getDomainFromImageName(chartSpec.Name)
+		if err != nil {
+			return ctrl.Result{}, errors.WithMessage(err, chartSpec.Name+": failed to parse image name")
+		}
+
+		creds, withCredentials := keyring.Lookup(chartSpec.Name)
+		if withCredentials {
+			for _, cred := range creds {
+				err := r.Helmer.RegistryLogin(repoToPull, cred.Username, cred.Password, false)
```
Comment (revit13) [inline manager/controllers/app/blueprint_controller.go:182]: Does RegistryLogout is also needed at the end?

Comment (cdoron) [reply manager/controllers/app/blueprint_controller.go:184]: good point. I will address this issue.

File: manager/controllers/app/blueprint_controller.go [hunk 3]
```diff
@@ -130,7 +135,16 @@ func (r *BlueprintReconciler) deleteExternalResources(blueprint *app.Blueprint)
 	return errors.New(strings.Join(errs, "; "))
 }
 
-func (r *BlueprintReconciler) applyChartResource(log logr.Logger, chartSpec app.ChartSpec, args map[string]interface{}, blueprint *app.Blueprint, releaseName string) (ctrl.Result, error) {
+func getDomainFromImageName(image string) (string, error) {
+	named, err := distributionref.ParseNormalizedNamed(image)
+	if err != nil {
+		return "", fmt.Errorf("couldn't parse image name: %v", err)
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:141]: Perhaps use errors.WithMessage for consistency?

File: manager/controllers/app/blueprint_controller.go [hunk 4]
```diff
@@ -141,7 +155,53 @@ func (r *BlueprintReconciler) applyChartResource(log logr.Logger, chartSpec app.
...
+
+		if err := r.Get(ctx, types.NamespacedName{Namespace: utils.GetSystemNamespace(), Name: chartSpec.ChartPullSecret},
+			&pullSecret); err == nil {
+			// if this is not a dockerconfigjson, ignore
+			if pullSecret.Type == "kubernetes.io/dockerconfigjson" {
+				pullSecrets = append(pullSecrets, pullSecret)
+			}
+		}
+
+		// create a keyring of all dockerconfigjson secrets, to be used for lookup
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:173]: At this point pullSecrets can be empty. You could try to pull chart anyway, but the credentialprovider functionality is not required in this case, right?

File: manager/controllers/app/blueprint_controller.go [hunk 5]
```diff
@@ -141,7 +155,53 @@ func (r *BlueprintReconciler) applyChartResource(log logr.Logger, chartSpec app.
...
+
+		creds, withCredentials := keyring.Lookup(chartSpec.Name)
+		if withCredentials {
+			for _, cred := range creds {
+				err := r.Helmer.RegistryLogin(repoToPull, cred.Username, cred.Password, false)
+				if err == nil {
+					registrySuccessfulLogin = repoToPull
+					break
+				} else {
+					log.Info("Failed to login to helm registry: " + repoToPull)
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:189]: Is this a real error? It's just one cred from a range. Perhaps to check for non-empty registrySuccessfulLogin right after the for loop?

File: manager/controllers/app/blueprint_controller.go [hunk 6]
```diff
@@ -141,7 +155,53 @@ func (r *BlueprintReconciler) applyChartResource(log logr.Logger, chartSpec app.
...
+				}
+			}
+		} else {
+			log.Info("Could not find ChartPullSecret. Will try to pull chart anyway")
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:193]: The printout is misleading. The error means that the required credentials were not supplied.

File: manager/controllers/app/blueprint_controller.go (hunk 4 above)
Comment (cdoron) [reply manager/controllers/app/blueprint_controller.go:173]: right. I will try to simplify

File: manager/controllers/app/blueprint_controller.go (hunk 5 above)
Comment (cdoron) [reply manager/controllers/app/blueprint_controller.go:189]: not an error, just an info message.

**Review by shlomitk1 — APPROVED** [review]

**Review by revit13 — APPROVED** [review]

**Pull request merged**